ch5/ex-5-13-tail.c

```diff
@@ -0,0 +1,225 @@
+ *
+ * 	tail -n
+ *
+ * prints the last n lines.  The program should behave rationally no matter
+ * how unreasonable the input or the value of n.  Write the program so it
+ * makes the best use of available storage; lines should be stored as in the
+ * sorting program of Section 5.6, not in a two-dimensional array of fixed
+ * size.
+ */
+
+#include <ctype.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#define INITIAL_BUFFSIZE (1 << 10)
+#define DEFAULT_TAILSIZE 10
+
+enum { OK, ERROR };
+
+/* Single buffer to store characters, loops back to start on reaching limit */
+unsigned long buffsize;
+unsigned long maxbuffsize = ULONG_MAX;
+char *buff, *cursor, *bufflimit;
+
+/* max number of lines to print */
+unsigned long tailsize;
+/* pointers to the start of lines in the buffer, loops back start at limit */
+char **lines, **firstln, **currln, **lineslimit;
+
+int readlines(void);
+void writelines(void);
+int init_buff(unsigned long);
+int init_lines(void);
+int enlargebuff(void);
+int settailsize(int argc, char **argv);
+
+int main(int argc, char **argv)
+{
+	if (settailsize(argc, argv) != OK)
+		return 0;
+	if (tailsize == 0)
+		return 0;
+	if (init_buff(INITIAL_BUFFSIZE) != OK)
+		return 0;
+	if (init_lines() != OK) {
+		free(buff);
+		return 0;
+	}
+
+	if (readlines() == OK)
+		writelines();
+
+	free(lines);
+	free(buff);
+	return 0;
+}
+
+int readlines(void)
+{
+	int c, prevc = '\0';
+
+	while ((c = getchar()) != EOF) {
+		if (prevc == '\n') {
+			if (++currln == lineslimit)
+				currln = lines;
+			*currln = cursor;
+			if (currln == firstln)
+				if (++firstln == lineslimit)
+					firstln = lines;
+		}
+		*cursor = prevc = c;
+		if (++cursor == bufflimit)
+			cursor = buff;
+		if (cursor == *firstln) {
+			if (enlargebuff() != OK)
+				return ERROR;
+		}
+	}
+	return OK;
+}
+
+void writelines(void)
+{
+	char *bptr;
+
+	if (*firstln <= cursor) {
+		for (bptr = *firstln; bptr < cursor; bptr++)
+			putchar(*bptr);
+	} else {
+		for (bptr = *firstln; bptr < bufflimit; bptr++)
+			putchar(*bptr);
+		for (bptr = buff; bptr < cursor; bptr++)
+			putchar(*bptr);
+	}
+}
+
+int init_buff(unsigned long size)
+{
+	buffsize = size;
+	buff = malloc(buffsize * sizeof(char));
+	if (buff == NULL) {
+		printf("error: Insufficient memory (allocating buff)\n");
+		return ERROR;
+	}
+	cursor = buff;
+	bufflimit = buff + buffsize;
+	return OK;
+}
+
+int init_lines(void)
+{
+	lines = malloc(tailsize * sizeof(char *));
+	if (lines == NULL) {
+		printf("error: Insufficient memory (allocating lines)\n");
+		return ERROR;
+	}
+	firstln = currln = lines;
+	*lines = buff;
+	lineslimit = lines + tailsize;
+	return OK;
+}
+
+int enlargebuff(void)
+{
+	unsigned long used;
+	char *bptr, **lptr;
+	ptrdiff_t offset;
+
+	/* copy info about current buffer before it's replaced */
+	char *pbuff = buff, *pcursor = cursor, *pbufflimit = bufflimit;
+	char *pfirstchr = *firstln;
+	ptrdiff_t pfirst2limit = pbufflimit - pfirstchr;
```

Comment:
Maybe consider have one pointer and then making the other variables `int` offsets instead.
I think it will make this function much simpler.

Reply:
Good point.  Orignally I thought I was saving a bunch of pointer arithmatic in the char loop.  But now you make me think about it...  if everything is an offset then most of the time offsets can be compared directly with each other.   The tiny bit of extra pointer arithmatic in the read char loop would, presumably, have virtually no overhead on a modern processor, but I do have one customer from Belgium who is already expressing concerns about performance :-)